File: modules/graphics/lib/rin/graphics/ImageBarrierOptions.cpp
#include "rin/graphics/ImageBarrierOptions.hpp"
ImageBarrierOptions& ImageBarrierOptions::DstAccess(vk::AccessFlags2 flags)
{
    dstAccessFlags = flags;
    return *this;
}

ImageBarrierOptions& ImageBarrierOptions::SrcAccess(vk::AccessFlags2 flags)
{
    srcAccessFlags = flags;
    return *this;
}

ImageBarrierOptions& ImageBarrierOptions::WaitStages(vk::PipelineStageFlags2 stages)
{
    waitForStages = stages;
    return *this;
}

ImageBarrierOptions& ImageBarrierOptions::NextStages(vk::PipelineStageFlags2 stages)
{
    nextStages = stages;
    return *this;
}

ImageBarrierOptions& ImageBarrierOptions::SubResource(const vk::ImageSubresourceRange& subresource)
{
    subresourceRange = subresource;
    return *this;
}

File: rin.Graphics/ImageBarrierOptions.cs
using TerraFX.Interop.Vulkan;

namespace rin.Graphics;

/// <summary>
///     Options for image barriers
/// </summary>
public struct ImageBarrierOptions()
{
    public VkAccessFlags2 DstAccessFlags =
        VkAccessFlags2.VK_ACCESS_2_MEMORY_WRITE_BIT | VkAccessFlags2.VK_ACCESS_2_MEMORY_READ_BIT;

    public VkPipelineStageFlags2 NextStages = VkPipelineStageFlags2.VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT;
    public VkAccessFlags2 SrcAccessFlags = VkAccessFlags2.VK_ACCESS_2_MEMORY_WRITE_BIT;
    public VkPipelineStageFlags2 WaitForStages = VkPipelineStageFlags2.VK_PIPELINE_STAGE_2_ALL_COMMANDS_BIT;

    public VkImageSubresourceRange SubresourceRange =
        SGraphicsModule.MakeImageSubresourceRange(VkImageAspectFlags.VK_IMAGE_ASPECT_COLOR_BIT);
    
}